src/mame/nintendo/playch10.cpp

```diff
@@ -310,38 +629,848 @@ WRITE_LINE_MEMBER(playch10_state::up8w_w)
+{
+	// Toggling bit 0 high then low resets both controllers
+	if (data & 1)
+		return;
+
+	// load up the latches
+	for (int i = 0; i < 2; i++)
+		m_input_latch[i] = m_in[i]->read();
+
+	// apply any masking from the BIOS
+	if (m_cntrl_mask)
+	{
+		// mask out select and start
+		m_input_latch[0] &= ~0x0c;
+	}
+}
+
+u8 playch10_state::pc10_in0_r()
+{
+	int ret = m_input_latch[0] & 1;
+
+	// shift
+	m_input_latch[0] >>= 1;
+
+	// some games expect bit 6 to be set because the last entry on the data bus shows up
+	// in the unused upper 3 bits, so typically a read from $4016 leaves 0x40 there.
+	ret |= 0x40;
+
+	return ret;
+}
+
+u8 playch10_state::pc10_in1_r()
+{
+	int ret = m_input_latch[1] & 1;
+
+	// shift
+	m_input_latch[1] >>= 1;
+
+	// do the gun thing
+	if (m_pc10_gun_controller)
+	{
+		int trigger = m_in[0]->read();
+
+		if (!m_sensor->detect_light(m_gunx->read(), m_guny->read()))
+			ret |= 0x08;
+
+		// now, add the trigger if not masked
+		if (!m_cntrl_mask)
+			ret |= (trigger & 2) << 3;
+	}
+
+	// some games expect bit 6 to be set because the last entry on the data bus shows up
+	// in the unused upper 3 bits, so typically a read from $4016 leaves 0x40 there.
+	ret |= 0x40;
+
+	return ret;
+}
+/*************************************
+ *
+ *  PPU External bus handlers
+ *
+ *************************************/
+
+void playch10_state::pc10_chr_w(offs_t offset, u8 data)
+{
+	int bank = BIT(offset, 10, 3);
+	if (m_chr_page[bank].writable)
+		m_chr_page[bank].chr[offset & 0x3ff] = data;
+}
+
+u8 playch10_state::pc10_chr_r(offs_t offset)
+{
+	int bank = BIT(offset, 10, 3);
+	return m_chr_page[bank].chr[offset & 0x3ff];
+}
+
+void playch10_state::pc10_set_mirroring(int mirroring)
+{
+	int bit = mirroring == PPU_MIRROR_HORZ || mirroring == PPU_MIRROR_HIGH;
+
+	switch (mirroring)
+	{
+		case PPU_MIRROR_LOW:
+		case PPU_MIRROR_HIGH:
+			for (int i = 0; i < 4; i++)
+				m_nt_page[i]->set_entry(bit);
+			break;
+		case PPU_MIRROR_HORZ:
+		case PPU_MIRROR_VERT:
+		case PPU_MIRROR_4SCREEN:
+		default:
+			for (int i = 0; i < 4; i++)
+				m_nt_page[i]->set_entry(BIT(i, bit));
+			break;
+	}
+}
+
+/* SIZE MAPPINGS *\
+ * old       new *
+ * 512         8 *
+ * 256         4 *
+ * 128         2 *
+ *  64         1 *
+\*****************/
+
+void playch10_state::pc10_set_videorom_bank(int first, int count, int bank, int size)
+{
+	// first = first bank to map
+	// count = number of 1K banks to map
+	// bank = index of the bank
+	// size = size of indexed banks (in KB)
+	// note that this follows the original PPU banking and might be overly complex
+
+	// yeah, this is probably a horrible assumption to make
+	// but the driver is 100% consistant
+
+	if (!m_vrom)    // do nothing if there is no CHR ROM to bank
+		return;
+
+	int len = m_vrom_region->bytes();
+	len /= 0x400;   // convert to KB
+	len /= size;    // convert to bank resolution
+	len--;          // convert to mask
+	bank &= len;    // should be the right mask
```

Comment:
This should have an assert that the length was actually a power of two – otherwise it’s too easy to get broken by someone adding a set where it’s a non power-of-two size.

Reply:
Asserts added to machine_start. The CHR banking is code I haven't really touched. FWIW, it seemed "good enough" for the moment and in the long run I was hoping to replace it all with proper slotted cart code.

Reply:
Yeah, it does need to be changed to use slots eventually, and the single-screen system needs to be turned into a separate machine configuration as it should be.